### PR TITLE
⚡ Bolt: Pre-allocate buffers in ShardedVectorIndex batch operations

### PR DESCRIPTION
**Pull request by madmax983**
⚡ Bolt: Pre-allocate buffers in ShardedVectorIndex batch operations

💡 **What:**
Optimized `add_batch_ref` and `remove_batch` in `src/index/vector/sharded.rs` to pre-allocate shard-specific vectors with estimated capacity.

🎯 **Why:**
The previous implementation used `vec![Vec::new(); self.shards.len()]`, which initializes vectors with 0 capacity. As items were distributed to shards, these vectors had to resize multiple times (e.g., 0 -> 4 -> 8 -> 16 -> ... -> 256), causing unnecessary heap allocations and data copying.

📊 **Impact:**
- Eliminates vector resizing during batch grouping.
- For a batch of N items and S shards, saves approximately `S * log2(N/S)` reallocations.
- Example: 1000 items, 4 shards -> Saves ~28 reallocations.

🔬 **Measurement:**
Verified using a reproduction script that tracked `Vec::capacity()` growth.
- **Before:** Capacity grew dynamically (e.g., 0 -> 256).
- **After:** Capacity remained constant at the estimated size (e.g., 251).

---
*PR created automatically by Jules for task [2059973877101976861](https://jules.google.com/task/2059973877101976861) started by @madmax983*

### CODE REVIEW

**Comment by madmax983** [comment]
> ## Code Review: Pre-allocate buffers in ShardedVectorIndex batch operations
> ### Summary
> This PR optimizes batch operations by pre-allocating vectors with estimated capacity instead of starting with zero capacity. The change is sound and addresses a real performance issue.
> 
> ### ✅ Strengths
> 1. **Well-motivated optimization**: The PR description clearly explains the problem (repeated reallocations) and quantifies the impact.
> 2. **Correct capacity calculation**: The formula `items.len() / self.shards.len() + 1` is appropriate for uniform hash-based distribution:
>    
>    * The division gives average items per shard
>    * The `+ 1` provides a small buffer for rounding
>    * Hash-based sharding (line 318-321) ensures near-uniform distribution
> 3. **Consistent implementation**: Both `add_batch_ref` and `remove_batch` use the same pattern.
> 4. **Good test coverage**: Existing tests cover the batch operations (lines 924-942, 1562-1576).
> 
> ### ⚠️ Potential Issues & Suggestions
> #### 1. **Edge Case: Division by Zero (CRITICAL)**
> **Issue**: If `self.shards.len()` is 0, the capacity calculation will panic.
> 
> **Location**: Lines 655, 682
> 
> ```rust
> let capacity = items.len() / self.shards.len() + 1;  // Panics if shards is empty
> ```
> 
> **Analysis**:
> 
> * The constructor clamps `num_shards` to at least 1 (need to verify)
> * There's a debug_assert at line 314: `debug_assert!(!self.shards.is_empty())`
> * However, debug_asserts are removed in release builds
> 
> **Recommendation**: Add a comment referencing the constructor invariant, or consider using `checked_div()` with a fallback:
> 
> ```rust
> let capacity = items.len().checked_div(self.shards.len()).unwrap_or(items.len()) + 1;
> ```
> 
> #### 2. **Edge Case: Empty Batch**
> **Current behavior**: When `items.len() == 0`, capacity becomes `1`.
> 
> **Analysis**: This is harmless but slightly wasteful. Consider:
> 
> ```rust
> let capacity = if items.is_empty() {
>     0
> } else {
>     items.len() / self.shards.len() + 1
> };
> ```
> 
> **Priority**: Low (1 extra usize allocation is negligible)
> 
> #### 3. **Worst-Case Distribution**
> **Issue**: In pathological cases where hash collisions cause all items to route to one shard, we over-allocate for `S-1` shards and under-allocate for 1 shard.
> 
> **Example**: 1000 items, 4 shards, all hash to shard 0:
> 
> * Each shard gets capacity 251
> * Shard 0 needs 1000, will resize anyway
> * Shards 1-3 waste 251 allocations each
> 
> **Mitigation**: Hash-based sharding makes this extremely unlikely. The current approach optimizes for the common case (uniform distribution).
> 
> **Recommendation**: No change needed, but could add a comment explaining the assumption.
> 
> #### 4. **Missing Benchmark**
> **Issue**: The PR claims performance improvements but doesn't add a benchmark to verify.
> 
> **Recommendation**: Add a benchmark in `benches/` to measure batch operation performance:
> 
> ```rust
> #[bench]
> fn bench_sharded_add_batch(b: &mut Bencher) {
>     let index = ShardedVectorIndex::with_defaults(4, DistanceMetric::Cosine, 128)?;
>     let items: Vec<_> = (0..1000)
>         .map(|i| (NodeId::new(i).unwrap(), vec![0.1; 128]))
>         .collect();
>     
>     b.iter(|| {
>         index.add_batch(&items)
>     });
> }
> ```
> 
> ### 🔍 Code Quality
> ✅ **Formatting**: Code follows Rust style guidelines ✅ **Comments**: Clear inline comments explain the optimization ✅ **Type Safety**: Uses proper newtypes (NodeId) ✅ **Error Handling**: Propagates errors correctly with `?` ✅ **Thread Safety**: No new thread safety concerns
> 
> ### 🔒 Security
> ✅ **No DoS vectors**: Pre-allocation is bounded by input size ✅ **No overflow risks**: Division is safe (assuming shards.len() > 0)
> 
> ### 📊 Performance
> **Expected Impact** (from PR description):
> 
> * Eliminates `S * log2(N/S)` reallocations
> * Example: 1000 items, 4 shards → saves ~28 reallocations
> 
> **Actual Impact**: Should be measurable in benchmarks, especially for large batches.
> 
> ### ✅ Recommended Actions Before Merge
> 1. **[OPTIONAL]** Add a comment explaining the uniform distribution assumption
> 2. **[RECOMMENDED]** Add benchmark to verify performance improvement
> 3. **[OPTIONAL]** Add test case for empty batch (`items.len() == 0`)
> 4. **[REQUIRED]** Verify clippy passes: `cargo clippy --all-targets --all-features -- -D warnings`
> 5. **[REQUIRED]** Verify tests pass: `cargo test`
> 
> ### Final Verdict
> **LGTM with minor suggestions** ✅
> 
> The core optimization is correct and valuable. The edge cases identified are either:
> 
> * Protected by constructor invariants (division by zero)
> * Negligible performance impact (empty batch)
> * Extremely unlikely (worst-case distribution)
> 
> This PR improves performance without introducing bugs or security issues. The suggestions above are optional improvements that could be addressed in follow-up PRs.
> 
> **Estimated Performance Gain**: 5-15% for batch sizes > 100, more significant for larger batches.

@jules

File: src/index/vector/sharded.rs [hunk 1]
```diff
@@ -651,7 +651,11 @@ impl VectorIndex for ShardedVectorIndex {
 
     fn add_batch_ref(&self, items: &[(NodeId, &[f32])]) -> Result<()> {
         // Group item indices by shard first
-        let mut shard_indices: Vec<Vec<usize>> = vec![Vec::new(); self.shards.len()];
+        // Pre-allocate to avoid resizing: assume uniform distribution
+        let capacity = items.len() / self.shards.len() + 1;
+        let mut shard_indices: Vec<Vec<usize>> = (0..self.shards.len())
+            .map(|_| Vec::with_capacity(capacity))
+            .collect();
```
[the remark this answers is not in the document]
Comment (madmax983) [reply src/index/vector/sharded.rs:667]: > `capacity = items.len() / self.shards.len() + 1` will allocate per-shard buffers even for an empty batch (capacity becomes 1), and it will panic with “divide by zero” if `self.shards` is ever empty (possible today via `ShardedVectorIndex::load` when `config.num_shards == 0`). Consider an early return for `items.is_empty()`, and compute capacity with a ceiling division that yields 0 for empty batches (and/or return a proper error if `self.shards.is_empty()` rather than panicking).

@jules

File: src/index/vector/sharded.rs [hunk 2]
```diff
@@ -674,7 +678,11 @@ impl VectorIndex for ShardedVectorIndex {
 
     fn remove_batch(&self, ids: &[NodeId]) -> Result<()> {
         // Group IDs by shard
-        let mut shard_ids: Vec<Vec<NodeId>> = vec![Vec::new(); self.shards.len()];
+        // Pre-allocate to avoid resizing: assume uniform distribution
+        let capacity = ids.len() / self.shards.len() + 1;
+        let mut shard_ids: Vec<Vec<NodeId>> = (0..self.shards.len())
+            .map(|_| Vec::with_capacity(capacity))
+            .collect();
```
[the remark this answers is not in the document]
Comment (madmax983) [reply src/index/vector/sharded.rs:703]: > Same issue here: `capacity = ids.len() / self.shards.len() + 1` forces allocations even when `ids` is empty, and will panic on divide-by-zero if `self.shards` is empty (which can currently happen via `load` when `config.num_shards == 0`). Consider early-returning on an empty slice and/or returning an explicit error when shard count is 0, and use a ceiling division that yields 0 for empty batches.

@jules
